### PR TITLE
feat(team-selector): Implement useTeams hook for searching

### PR DESCRIPTION
**Pull request by evanpurkhiser**
This react hook allows us to easily implement type-ahead for teams.

In the future this will also support loading specific teams, and loading teams that you're a part of. We may potentially even add support to incrementally load results via pagination.

### CODE REVIEW

File: static/app/components/forms/teamSelector.tsx [hunk 1]
```diff
@@ -249,12 +251,14 @@ function TeamSelector(props: Props) {
     <SelectControl
       ref={selectRef}
       options={options}
+      onInputChange={debounce(val => void onSearch(val), DEFAULT_DEBOUNCE_DURATION)}
```
Comment (davidenwang) [inline static/app/components/forms/teamSelector.tsx:254]: is the void here just for typing or does debounce do something with the return value of the debounced func?

File: static/app/utils/useTeams.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+};
+
+export type Result = {
+  /**
+   * The loaded teams list
+   */
+  teams: Team[];
+  /**
+   * This is an action provided to consumers for them to update the current
+   * teams result set using a simple search query. You can allow the new
```
Comment (davidenwang) [inline static/app/utils/useTeams.tsx:45]: this comment can be changed to remove the bit about appending/replacing

File: tests/js/spec/components/forms/teamSelector.spec.jsx [hunk 3]
```diff
@@ -26,16 +27,11 @@ const teamData = [
 ];
 const teams = teamData.map(data => TestStubs.Team(data));
 const project = TestStubs.Project({teams: [teams[0]]});
+const organization = TestStubs.Organization({access: ['project:write']});
```
Comment (davidenwang) [inline tests/js/spec/components/forms/teamSelector.spec.jsx:30]: oh shoot thanks for this, looks much better

File: tests/js/spec/utils/useTeams.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    await onSearchPromise;
+    expect(result.current.fetching).toBe(false);
+
+    // State will be updated due to store chage
+    await waitForNextUpdate();
+
+    expect(mockRequest).toHaveBeenCalled();
+    expect(result.current.teams).toEqual([...mockTeams, newTeam2, newTeam3]);
+
+    // de-duplicates itesm in the query results
```
Comment (davidenwang) [inline tests/js/spec/utils/useTeams.spec.tsx:56]: nit: items*

**Review by davidenwang — APPROVED** [review]
✨ HOOKS 🎣 

nice work im excited for this

File: static/app/components/forms/teamSelector.tsx (hunk 1 above)
Comment (EvanPurkhiser) [reply static/app/components/forms/teamSelector.tsx:254]: yeah the onInputChange will actually try and render what is returned here, hence the void.

I could also write as `{ ... }`

File: static/app/utils/useTeams.tsx (hunk 2 above)
Comment (EvanPurkhiser) [reply static/app/utils/useTeams.tsx:45]: Ah yea

File: tests/js/spec/components/forms/teamSelector.spec.jsx (hunk 3 above)
Comment (EvanPurkhiser) [reply tests/js/spec/components/forms/teamSelector.spec.jsx:30]: needed to update it since it doesn't take those as props now :)

**Pull request merged**